src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -864,7 +864,36 @@ internal static ImmutableArray<T> AddRange<T>(this ImmutableArray<T> self, in Te
             return builder.ToImmutableAndFree();
         }
 
-        internal static bool HasDuplicates<T>(this ImmutableArray<T> array, IEqualityComparer<T>? comparer = null)
+        internal static bool HasDuplicates<T>(this ImmutableArray<T> array)
+        {
+            switch (array.Length)
+            {
+                case 0:
+                case 1:
+                    return false;
+
+                case 2:
+                    return EqualityComparer<T>.Default.Equals(array[0], array[1]);
+
+                default:
+                    var set = PooledHashSet<T>.GetInstance();
+                    var foundDuplicate = false;
+
+                    foreach (var element in array)
+                    {
+                        if (!set.Add(element))
+                        {
+                            foundDuplicate = true;
+                            break;
+                        }
+                    }
+
+                    set.Free();
+                    return foundDuplicate;
+            }
+        }
+
+        internal static bool HasDuplicates<T>(this ImmutableArray<T> array, IEqualityComparer<T> comparer)
```

Comment:
nit, have this call into the other if comparer is null or EqualityComparer<T>.Default.

Reply:
I'm going to hold off on making that change unless you or someone else feels strongly for it. My thought it that it adds a bit of complication to this code to better improve perf for a caller that simply shouldn't pass in the 2nd param.

Reply:
Agreed with Todd here. It doesn't seem worth the code complexity.